---
title: "`ector`: Ectopic gene expression using R"
subtitle: "TCGA-COAD case study"
author: "MÃ©lanie Petrier, Ekaterina Flin, Florent Chuffart, Sophie Rousseaux"
contact: 
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r echo=FALSE, eval=TRUE}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=TRUE, results="verbatim", dpi=75)
```





# Introduction

ECTOR is a R package that detects ectopic gene expression in cancer, according to a threshold determined from normal samples. 
ECTOR analyses then the impact of abnormal expression of these genes in patient survival, allowing the detection of potential biomarkers of cancer aggressiveness.



# Dataset


The illustration of the method required the import of transcriptomic data (RNAseq) of 25.360 genes for 499 samples and clinical data (experimental grouping) of 547 samples from the TCGA study of colon adenocarcinoma. The list of tissue-restricted genes (testis, placenta, embryonic stem cells and brain) has also been imported from the Epimed database.

`e` is a data matrix containing clinical information for 547 samples. The columns represent these clinical characteristics and each row corresponds to one patient, whose ID is given in the first column (id_sample).

```{r}
load("~/projects/ector/data/ector_clinicals.RData")
e = ector_clinicals
head(e[,1:4])
dim(e)
```
 
 
`d`  is a data matrix containing the RNAseq counts of each gene for 41 control samples and 458 tumorous samples, meaning 499 samples in total. It has been obtain from the TCGA study of colon adenocarcinoma (`s$data`). The rownames and colnames represent respectively the gene symbols and sample ID.

```{r}
load("~/projects/ector/data/ector_transcriptome.RData")
d = ector_transcriptome
head(d[,1:4])
dim(d)
```


`idx_n` indicates the ID of normal samples and `idx_c` indicates the ID of cancerous samples.

```{r}
idx_n = rownames(e)[e$tissue_status %in% "normal"]  
head(idx_n)
length(idx_n)
idx_c = rownames(e)[e$tissue_status %in% "tumoral"]
head(idx_c)
length(idx_c)
``` 
 


# Method

## Establishment of gene expression threshold

In order to know if genes are abnormally expressed in cancerous samples, it is required to establish a threshold from normal samples for all genes. The formula commonly used is mean(x) + 3*sd(x), x being the expression of a given gene in normal samples.
`m3sd` is the function used to determine the threshold for each gene and uses only data from normal samples (`idx_n`). The function is then applied for each gene in the data matrix `d`.

```{r}
m3sd = function(x) {
  mean(x)+(3*sd(x))
}

thresh = apply(d[,idx_n], 1, m3sd)
head(thresh)
length(thresh)
```



## Plotting gene expression

It is possible to visualize approximately the number of samples where a given gene is abnormally expressed.
`figure` is a function to draw a plot where each sample is represented by a dot. The x axis is the tissue status (normal or tumoral) and the y axis represents the expression. 
The plot contains also the lines corresponding to the mean and the threshold (m3sd) of the given gene and the dots found above the threshold are colored in red.


```{r}
plot_gene_expr = function(gs, d, idx_n, thresh_func=m3sd) {
  expr = d[gs,]
  plot(jitter(as.numeric(as.factor(e$tissue_status))), expr, 
  xlab=" ", ylab="Expression", xaxt = "n",
       col = (expr>thresh_func(expr[idx_n]))+1, 
       main=gs
  )
  tmp_tab = table(e$tissue_status)
  axis(1, at = 1:2, paste0(names(tmp_tab)," (n=", tmp_tab, ")"), las = 1)
  abline(h=thresh_func(expr[idx_n]), col="red")
  abline(h=mean(expr[idx_n]), col="black")
  legend("topleft", c("mean", "threshold"), col=1:2, lty=1)
}

```



## Determination of ectopically expressed genes

In order to know if a gene is ectopically expressed in cancerous samples, it is required to compare their expression with the corresponding threshold.
`prop_aberrant_expr` is a function which aims to calculate the percentage of patients in which the expression of a given gene is aberrant (higher than the threshold).
`candidates` indicates the gene symbols that have an aberrant expression in at least 10% of patients. 


```{r}
prop_ectopic_expr = function(x, idx_c, idx_n, thresh_func=m3sd) {
  if (sum(is.na(x)) > 0) {
    warning("NA is expression values.")    
  }
  y = x[idx_c]>thresh_func(x[idx_n])
  res = sum(y)/length(idx_c)
  return(res)
}

ectopic_activation = apply(d, 1, prop_ectopic_expr, idx_c, idx_n)
head(ectopic_activation)
length(ectopic_activation)
candidates = names(ectopic_activation)[ectopic_activation>0.1]
length(candidates)

```


## Survival

### p-values

The influence of ectopic gene expression on patient survival is determined by the calculation of cox p-values by using the function `compute_cox_pv`.
`pvs` allows to apply this function to every candidate gene.
`biomarkers` gives the gene symbol for each gene whose cox p-value is less than 0.05.


```{r}
compute_cox_pv = function (x, idx_c, idx_n, e, thresh_func=m3sd) {
  ss = survival::Surv(e[idx_c,]$os_months, e[idx_c,]$os_censor)
  v = x[idx_c]>thresh_func(x[idx_n])
  f = suppressWarnings(survival::coxph(ss ~ v))
  sf = summary(f)
  pvcox = signif(sf$wald["pvalue"], digits=3)
  hr = signif(sf$coef[2], digits=3)
  logrank = signif(sf$logtest["pvalue"], digits=3)
  res = c(pvcox, hr, logrank)
  return(res)
  
}

pvs = apply(d[candidates,], 1, compute_cox_pv, idx_c, idx_n, e)
head(pvs)

pvcox = pvs[1,]
biomarkers = names(pvcox)[pvcox<0.05]
pvs_biomarkers = apply(d[biomarkers,], 1, compute_cox_pv, idx_c, idx_n, e)
hr = pvs_biomarkers[2,]
biomarkers = names(hr)[hr>1]
logrank = pvs[3,]

```


### Curves

As for p-values, the function drawing survival curves (`plot_survival`) requires time and censor of survival. Patients are divided in two groups, depending on the activation status of a given gene (on or off). `plot_survival` is then applied to the previously determined `biomarkers`.


```{r}

plot_survival = function(gs, d, idx_c, idx_n, e, thresh_func=m3sd) {
  expr = d[gs,]
  pv = compute_cox_pv(expr, idx_c, idx_n, e, thresh_func=m3sd)
  ss = survival::Surv(e[idx_c,]$os_months, e[idx_c,]$os_censor)
  v = expr[idx_c]>thresh_func(expr[idx_n])
  sf = survival::survfit(ss ~ v)
  plot(
    sf, 
    xlab="Time in months", 
    ylab="Overall survival", 
    main=paste0(gs, " pv_cox=", signif(pvcox[gs], 3),
                "\n pv_logrank=", signif(logrank[gs],3),
                "\n HR=", signif(hr[gs],3)),
    col=c(4,2)
  )
  legend("topright",
        c(paste0("Off (n=", sum(expr[idx_c]<=thresh_func(expr[idx_n])), ")"),
        paste0("On (n=", sum(expr[idx_c]>thresh_func(expr[idx_n])), ")")),
  lty=1, col=c(4,2))
}

plot_survival(gs=rownames(d)[1], d, idx_c, idx_n, e, thresh_func=m3sd)

sapply(biomarkers, plot_survival, d, idx_c, idx_n, e)
```


# Biomarkers

```{r}
ector_biomarkers = biomarkers
save(ector_biomarkers, file='~/projects/ector/data/ector_biomarkers.RData' , compress='xz')
```


# Material and methods







# Session Information

```{r, results="verbatim"}
sessionInfo()
```



